imagej/imagej.py

```diff
@@ -286,19 +294,159 @@ def run_script(self, language, script, args=None):
+            """Convert the IJ uppercase dimension convention (X, Y, Z< C, T) to lowercase python (x, y, z, c, t) """
+            if str(axis) in ['X', 'Y', 'Z', 'C', 'T']:
+                return str(axis).lower()
+            return str(axis)
+
+        def _get_axis_num(self, xarr, axis):
+            """
+            Get the xarray -> java axis number due to inverted axis order for C style numpy arrays (default)
+            :param xarr: Xarray to convert
+            :param axis: Axis number to convert
+            :return: Axis idx in java
+            """
+            py_axnum = xarr.get_axis_num(axis)
+            if numpy.isfortran(xarr.values):
+                return py_axnum
+
+            return len(xarr.dims) - py_axnum - 1
+
+
+        def _assign_dataset_metadata(self, dataset, attrs):
+            """
+            :param dataset: ImageJ Java dataset
+            :param attrs: Dictionary containing metadata
+            """
+            dataset.getProperties().putAll(self.to_java(attrs))
+
+        def _axis_is_linear(self, coords):
+            """
+            Check if each axis has linear steps between grid points.  Skip over axes with non-numeric entries
+            :param coords: Xarray coords variable, which is a dict with axis: [axis values]
+            :return: Whether all axes are linear, or not.
+            """
+            linear = True
+            for coord, values in coords.items():
+                try:
+                    diff = numpy.diff(coords)
+                    if len(numpy.unique(diff)) > 1:
+                        warnings.warn(f'Axis {coord} is not linear')
+                        linear = False
+                except TypeError:
+                    continue
+            return linear
+
+        def _get_origin(self, axis):
+            """
+            Get the coordinate origin of an axis, assuming it is the first entry.
+            :param axis: A 1D list like entry accessible with indexing, which contains the axis coordinates
+            :return: The origin for this axis.
+            """
+            return axis.values[0]
+
+        def _get_scale(self, axis):
+            """
+            Get the scale of an axis, assuming it is linear and so the scale is simply second - first coordinate.
+            :param axis: A 1D list like entry accessible with indexing, which contains the axis coordinates
+            :return: The scale for this axis or None if it is a non-numeric scale.
+            """
+            try:
+                return axis.values[1] - axis.values[0]
+            except TypeError:
+                return None
+
+        def _java_to_dataset(self, data):
             """
             Converts the data into a ImageJ Dataset
             """
             try:
                 if self._ij.convert().supports(data, Dataset):
                     return self._ij.convert().convert(data, Dataset)
+                if self._ij.convert().supports(data, ImgPlus):
```

Comment:
Let's add an `NB: ` comment before all these checks, to explain why we do this craziness. It's really just that the set of converters is currently incomplete, and being thorough like this ensures we can get where we need to be. Yeah?